wiregrid: Add bias step measurements during calibration

sadachi5 opened the pull request:
### Purpose
To understand the time constant deeply, we would like to perform bias step measurements before and after the wiregrid calibration. 

### Changes
- Add bias step measurements before and after the `insert()` and  also `eject()` in `calibrate()` function.
- Add exception for `insert()` in the `calibrate()` function, and add `eject()` in the exception.

### Changed files
- `src/sorunlib/wiregrid.py`
- `tests/test_wiregrid.py`

BrianJKoopman commented:
> I'm not sure how to solve codecov error, we may need Brian's advice.

If you click through on the failing check, '...' -> 'View details' -> 'View this Pull Request on Codecov', you'll see a page like this: https://app.codecov.io/gh/simonsobs/sorunlib/pull/239?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=simonsobs

There you'll see lines not covered by tests. The decrease in test coverage is what's causing the check to fail. So the test cases do not cover, in this instance, the lines where an insert fails and we try to eject.

I'm wondering about the utility of ejecting after a failed insert. How likely is it that this works? Everywhere else we just fail when an insert/eject happens and state that the position needs to be inspected before continuing observations. I sort of think we should do the same here.

sadachi5 commented:
> > > 1. I'm fine but I wonder having biasstep 4 times might be too much, we might only need twice.
> > 
> > 
> > I agree with you. I modified the argument. Now, we have `bias_step_wo_wg`(defualt=True) and `bias_step_wt_wg`(default=False). Nominally, we just perform the `bias_step` before insert and after eject (without wiregrid on the window). But, for the test or future request, I added the argument `bias_step_wt_wg`, which will perform the bias_step before and after the rotation with wiregrid on the window.
> 
> I should have commented more explicitly. I think bias_step_wt_wg is more important for wiregrid calibration so it should be run by default, bias_step_wo_wg is helpful to understand instruments but it's less useful for calibration. I was actually thinking in different way. I thought running pair of biasstep with and without wiregrid before "calibration" might be enough. because the duration of wg measurement is short and drift of detector properties are likely to be small. What do you think?

Oh, I misunderstood. I agree that we prefer to have two bias steps with and without wg before calibration. For the time drift, the stepwise rotation itself takes ~4min. Probably, there is not much change. I changed the arguments to `bias_step_before` and `bias_step_after`, and set `before` as True and `after` as False for testing, just in case we want to do that.

ykyohei commented on src/sorunlib/wiregrid.py:
```diff
@@ -288,38 +288,109 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
+
+        finally:
+            # Eject the wiregrid
+            eject()
```
I think we would like to delete this based on the discussion above.

ykyohei reviewed (commented):
Thank you for the changes, looks mostly good to me. I will leave Brian to approve this.

sadachi5 commented:
@BrianJKoopman 
I and Kyohei approved the current scripts. Please check and approve it to merge.

BrianJKoopman commented on src/sorunlib/wiregrid.py:
```diff
@@ -288,38 +288,109 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
+
+        finally:
+            # Stop SMuRF streams
+            stop_smurfs()
```
This whole block could be written in ~half the lines:

```suggestion
    try:
        # Bias step (before calibration without wiregrid) + Enable SMuRF streams
        if bias_step_before:
            run.smurf.bias_step(tag=f'wiregrid, wg_before_wo_wg{el_tag}', concurrent=True)
            time.sleep(5)
            run.smurf.stream('on', tag=f'wiregrid, wg_inserting{el_tag}', subtype='cal')
        else:
            run.smurf.stream('on', tag=f'wiregrid, {rotation_tag}{el_tag}', subtype='cal')

        # Insert the wiregrid
        insert()

        # Bias step (before calibration with wiregrid) + Enable SMuRF streams
        if bias_step_before:
            stop_smurfs()
            run.smurf.bias_step(tag=f'wiregrid, wg_before_wt_wg{el_tag}', concurrent=True)
            time.sleep(5)
            run.smurf.stream('on', tag=f'wiregrid, {rotation_tag}{el_tag}', subtype='cal')

        # Rotate the wiregrid
        rotate(continuous)

        # Bias step (before calibration with wiregrid) + Enable SMuRF streams
        if bias_step_after:
            stop_smurfs()
            run.smurf.bias_step(tag=f'wiregrid, wg_after_wt_wg{el_tag}', concurrent=True)
            time.sleep(5)
            run.smurf.stream('on', tag=f'wiregrid, wg_ejecting{el_tag}', subtype='cal')

        # Eject the wiregrid
        eject()

    finally:
        # Stop SMuRF streams
        stop_smurfs()

    if bias_step_after:
        # Bias step (after calibration without wiregrid)
        run.smurf.bias_step(tag=f'wiregrid, wg_after_wo_wg{el_tag}', concurrent=True)
        time.sleep(5)
```

The main calibration is these steps:
1. Turn on streams
2. Insert
3. Rotate
4. Eject
5. Turn off streams

This keeps that ordering, inserting bias steps + stream off/on commands as needed while we go through the steps. The additional nested `try/finally`'s for each start/stop of the streams aren't necessary if we wrap the whole thing in a `try`, with the final `stop_smurfs()` being in the associated `finally` block.

sadachi5 replied:
@BrianJKoopman Thank you for changing the structure to be more compact.
I have one question about the current structure.
If the `bias_step` fails, can `stop_smurfs()` function in `finally` work without any error? (Does the function skip to stop the smurf if there is no running stream?)

sadachi5 replied:
If `stop_smurfs()` only produces an extra error message, that's fine. Let's keep your modified version for now since it makes the script more readable.

sadachi5 commented:
> > @BrianJKoopman, @ykyohei I added comments on Brian's comments. Please check them.
> 
> Sorry, I'm not seeing responses or a new commit. Did your push not go through?

I did not push anything. I commented on each of your comments before. Please check them.

BrianJKoopman commented on src/sorunlib/wiregrid.py:
```diff
@@ -288,38 +288,109 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
+            error = "Wiregrid calibration has failed " + \
+                "due to a failure in inserting the wiregrid.\n" + str(e)
+            # Raise error and end the calibration
+            raise RuntimeError(error)
```
The Traceback will point to the line that caused the exception, so it should be clear from that where the exception was raised. In nextline this appears at the top of the page, rather than in the stdout at the bottom.

sadachi5 replied:
I removed the exception.

BrianJKoopman commented on src/sorunlib/wiregrid.py:
```diff
@@ -288,38 +288,109 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
+        if bias_step_before:
+            # Bias step (before calibration without wiregrid)
+            run.smurf.bias_step(tag=f'wiregrid, wg_before_wo_wg{el_tag}', concurrent=True)
+            time.sleep(5)
```
Fair enough, @ykyohei, feel free to still weigh in, but since it was what we're doing in the `time_constant()` function already, I'm fine with this.
